src/Tasks/Microsoft.NET.Build.Tasks/ResourceUpdater.cs

```diff
@@ -0,0 +1,320 @@
+            }
+
+            if (!UpdateResource(hUpdate, lpType, lpName, LangID_LangNeutral_SublangNeutral, data, (uint)data.Length))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return this;
+        }
+
+        /// <summary>
+        /// Write the pending resource updates to the target PE
+        /// file. After this, the ResourceUpdater no longer maintains
+        /// an update handle, and can not be used for further updates.
+        /// Throws an InvalidOperationException if Update() was already called.
+        /// </summary>
+        public void Update()
+        {
+            if (hUpdate == IntPtr.Zero)
+            {
+                ThrowExceptionForInvalidUpdate();
+            }
+
+            try
+            {
+                if (!EndUpdateResource(hUpdate, false))
+                {
+                    ThrowExceptionForLastWin32Error();
+                }
+            }
+            finally
+            {
+                hUpdate = IntPtr.Zero;
+            }
+        }
+
+
+        private bool EnumTypesCallback(IntPtr hModule, IntPtr lpType, IntPtr lParam)
+        {
+            var enumNamesCallback = new EnumResNameProc(EnumNamesCallback);
+            if (!EnumResourceNames(hModule, lpType, enumNamesCallback, lParam))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return true;
+        }
+
+        private bool EnumNamesCallback(IntPtr hModule, IntPtr lpType, IntPtr lpName, IntPtr lParam)
+        {
+            var enumLanguagesCallback = new EnumResLangProc(EnumLanguagesCallback);
+            if (!EnumResourceLanguages(hModule, lpType, lpName, enumLanguagesCallback, lParam))
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            return true;
+        }
+
+        private bool EnumLanguagesCallback(IntPtr hModule, IntPtr lpType, IntPtr lpName, ushort wLang, IntPtr lParam)
+        {
+            IntPtr hResource = FindResourceEx(hModule, lpType, lpName, wLang);
+            if (hResource == IntPtr.Zero)
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            // hResourceLoadedq is just a handle to the resource,
+            // which can be used to get the resource data
+            IntPtr hResourceLoaded = LoadResource(hModule, hResource);
+            if (hResourceLoaded == IntPtr.Zero)
+            {
+                ThrowExceptionForLastWin32Error();
+            }
+
+            // This doesn't actually lock memory. It just retrieves a
+            // pointer to the resource data. The pointer is valid
+            // until the module is unloaded.
+            IntPtr lpResourceData = LockResource(hResourceLoaded);
+            if (lpResourceData == IntPtr.Zero)
+            {
+                throw new Exception("Failed to lock resource.");
```

Comment:
Please pick a more specific type than System.Exception

Reply:
Introduced ResourceNotAvailableException for this.